File: elixir/worg-agent/lib/worg_agent/tool_registry.ex
defmodule WorgAgent.ToolRegistry do
  @moduledoc """
  Looks up tool implementations by name. Tools are registered via
  `config :worg_agent, :tools, [...]` in `config/config.exs`.

  The Loop (wb-nlln.21.5) consults this registry to (a) build the
  tool-use schema sent to the LLM and (b) dispatch tool_call results
  by name back into the right implementation.

  Extending: consumer apps (Watershed, third parties) add their own
  tool modules to the config list. Order doesn't matter — lookup is
  by `name/0` not list position.
  """

  @doc """
  Return the configured list of tool modules. Defaults to the built-in
  set (Bash, Read, Write, LuaEval) if no config is set.
  """
  @spec all() :: [module()]
  def all do
    Application.get_env(:worg_agent, :tools, [
      WorgAgent.Tools.Bash,
      WorgAgent.Tools.Read,
      WorgAgent.Tools.Write,
      WorgAgent.Tools.LuaEval
    ])
  end

  @doc """
  Look up a tool module by its canonical `name/0` value.
  Returns `nil` if no tool matches.
  """
  @spec lookup(String.t()) :: module() | nil
  def lookup(name) when is_binary(name) do
    Enum.find(all(), fn module -> module.name() == name end)
  end

  @doc """
  Build the tool-use catalog: a list of maps with `name`,
  `description`, and `input_schema`. The Llm client (wb-nlln.21.4)
  converts this into provider-specific tool-use payloads (OpenAI's
  `tools.function` shape, Anthropic's bare-tools shape, etc.).
  """
  @spec catalog() :: [map()]
  def catalog do
    Enum.map(all(), fn module ->
      %{
        "name" => module.name(),
        "description" => module.description(),
        "input_schema" => module.input_schema()
      }
    end)
  end

  @doc """
  Dispatch a tool call by name. Returns whatever the tool's
  `execute/2` returned, or `{:error, {:unknown_tool, name}}` if no
  matching tool is registered.
  """
  @spec dispatch(String.t(), map(), WorgAgent.Tool.ctx()) ::
          {:ok, String.t()} | {:error, term}
  def dispatch(name, args, ctx) do
    case lookup(name) do
      nil -> {:error, {:unknown_tool, name}}
      module -> module.execute(args, ctx)
    end
  end
end

File: crates/worg-agent/src/tool_registry.rs
//! Looks up tool implementations by name. Mirrors Elixir's
//! `WorgAgent.ToolRegistry`.
//!
//! Unlike the Elixir version (which reads `:tools` from
//! `Application.get_env`), the Rust registry is constructed
//! explicitly by the binary entry point. This trades a bit of
//! ergonomics for explicit dependency injection — useful in tests
//! and when embedding the runtime in another Rust process.

use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;

use crate::tool::{Tool, ToolError};
use crate::types::{ToolCtx, ToolOutput};

/// A name → tool map. Cheap to clone (uses `Arc` internally).
#[derive(Clone, Default)]
pub struct ToolRegistry {
    by_name: HashMap<&'static str, Arc<dyn Tool>>,
}

impl ToolRegistry {
    pub fn new() -> Self {
        Self {
            by_name: HashMap::new(),
        }
    }

    /// Register a tool. Last-registration-wins for duplicate names;
    /// debug builds panic on duplicates to catch wiring bugs early.
    pub fn register<T: Tool + 'static>(&mut self, tool: T) -> &mut Self {
        let name = tool.name();
        if cfg!(debug_assertions) && self.by_name.contains_key(name) {
            panic!(
                "tool {name} registered twice — registry construction order is wrong"
            );
        }
        self.by_name.insert(name, Arc::new(tool));
        self
    }

    /// Look up by canonical name. `None` if not registered.
    pub fn lookup(&self, name: &str) -> Option<Arc<dyn Tool>> {
        self.by_name.get(name).cloned()
    }

    /// Build the tool-use catalog: a list of `{name, description,
    /// input_schema}` objects. The LLM client converts this into the
    /// provider-specific tool-use payload (OpenAI's
    /// `tools[].function`, Anthropic's bare-tools shape).
    pub fn catalog(&self) -> Vec<Value> {
        self.by_name
            .values()
            .map(|t| {
                serde_json::json!({
                    "name": t.name(),
                    "description": t.description(),
                    "input_schema": t.input_schema()
                })
            })
            .collect()
    }

    /// Dispatch a tool call by name. Returns
    /// `Err(DispatchError::UnknownTool)` if no matching tool is
    /// registered, otherwise propagates the tool's result.
    pub async fn dispatch(
        &self,
        name: &str,
        args: Value,
        ctx: &ToolCtx,
    ) -> Result<ToolOutput, DispatchError> {
        let tool = self
            .lookup(name)
            .ok_or_else(|| DispatchError::UnknownTool(name.to_string()))?;
        tool.execute(args, ctx)
            .await
            .map_err(DispatchError::ToolFailed)
    }

    pub fn len(&self) -> usize {
        self.by_name.len()
    }

    pub fn is_empty(&self) -> bool {
        self.by_name.is_empty()
    }

    pub fn names(&self) -> impl Iterator<Item = &'static str> + '_ {
        self.by_name.keys().copied()
    }
}

#[derive(Debug, thiserror::Error)]
pub enum DispatchError {
    #[error("unknown tool: {0}")]
    UnknownTool(String),
    #[error(transparent)]
    ToolFailed(#[from] ToolError),
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_trait::async_trait;

    struct NoopTool;

    #[async_trait]
    impl Tool for NoopTool {
        fn name(&self) -> &'static str {
            "noop"
        }
        fn description(&self) -> &'static str {
            "does nothing"
        }
        fn input_schema(&self) -> Value {
            serde_json::json!({"type": "object"})
        }
        async fn execute(&self, _args: Value, _ctx: &ToolCtx) -> Result<ToolOutput, ToolError> {
            Ok("ok".into())
        }
    }

    #[tokio::test]
    async fn unknown_tool_returns_dispatch_error() {
        let registry = ToolRegistry::new();
        let ctx = ToolCtx::sandboxed("/tmp");
        let err = registry
            .dispatch("missing", serde_json::json!({}), &ctx)
            .await
            .unwrap_err();
        match err {
            DispatchError::UnknownTool(name) => assert_eq!(name, "missing"),
            other => panic!("expected UnknownTool, got {other:?}"),
        }
    }

    #[tokio::test]
    async fn registered_tool_dispatches_through_registry() {
        let mut registry = ToolRegistry::new();
        registry.register(NoopTool);
        let ctx = ToolCtx::sandboxed("/tmp");
        let out = registry
            .dispatch("noop", serde_json::json!({}), &ctx)
            .await
            .unwrap();
        assert!(matches!(out, ToolOutput::Text(ref t) if t == "ok"));
    }

    #[test]
    fn catalog_contains_one_entry_per_registered_tool() {
        let mut registry = ToolRegistry::new();
        registry.register(NoopTool);
        let cat = registry.catalog();
        assert_eq!(cat.len(), 1);
        assert_eq!(cat[0]["name"], "noop");
    }
}
